hpx/util/unwrap.hpp

```diff
@@ -0,0 +1,166 @@
+//  Copyright (c) 2017 Denis Blank
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_UNWRAP_HPP
+#define HPX_UTIL_UNWRAP_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/util/detail/unwrap_impl.hpp>
+
+#include <cstddef>
+#include <utility>
+
+namespace hpx {
+namespace util {
+    /// A helper function for retrieving the actual result of
+    /// any hpx::lcos::future like type which is wrapped in an arbitrary way.
+    ///
+    /// Unwraps the given pack of arguments, so that any hpx::lcos::future
+    /// object is replaced by its future result type in the argument pack:
+    /// - `hpx::future<int>` -> `int`
+    /// - `hpx::future<std::vector<float>>` -> `std::vector<float>`
+    /// - `std::vector<future<float>>` -> `std::vector<float>`
+    ///
```

Comment:
That's great. We just ran into the need for `when_all()` to be able to handle a `vector<vector<future<T>>>` the other day...